File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/repository/retrofit/service/RetrofitGroupService.java
package app.android.ttp.mikazuki.yoshinani.repository.retrofit.service;

import android.support.annotation.NonNull;

import com.google.common.collect.Lists;

import java.util.List;

import app.android.ttp.mikazuki.yoshinani.model.GroupModel;
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.entity.Group;
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.entity.GroupUser;
import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import rx.Observable;

/**
 * @author haijimakazuki
 */
public interface RetrofitGroupService {

    static final String PATH_GROUPS = "groups";
    static final String PATH_GROUP_WITH_ID = "groups/{id}";
    static final String PATH_GROUP_USER = "groups/{group_id}/users";
    static final String PATH_GROUP_USER_ACCEPT = "groups/{group_id}/users/accept";

    @GET(PATH_GROUPS)
    public Observable<Response<List<Group>>> getAll(@Query("user_id") int userId);

    @GET(PATH_GROUP_WITH_ID)
    public Observable<Response<Group>> get(@Path("id") int groupId);

    @POST(PATH_GROUPS)
    public Observable<Response<Group>> create(@Body RequestDataOnCreate data);

    @PATCH(PATH_GROUP_WITH_ID)
    public Observable<Response<Group>> update(@Path("id") int groupId, @Body RequestDataOnCreate data);

    @POST(PATH_GROUP_USER)
    public Observable<Response<Group>> invite(@Path("group_id") int groupId,
                                              @Body RequestDataOnInvite data);

    @PATCH(PATH_GROUP_USER_ACCEPT)
    public Observable<Response<GroupUser>> accept(@Path("group_id") int groupId);

    public class RequestDataOnCreate {
        public PostData group;

        public RequestDataOnCreate(@NonNull final GroupModel group) {
            this.group = new PostData(group.createEntity());
        }

        class PostData {
            public String name;
            public String description;

            public PostData(@NonNull final Group group) {
                this.name = group.getName();
                this.description = group.getDescription();
            }
        }
    }

    public class RequestDataOnInvite {
        public List<PostData> groupUser;

        public RequestDataOnInvite(@NonNull final int userId) {
            this.groupUser = Lists.newArrayList(new PostData(userId));
        }

        class PostData {
            public int userId;

            public PostData(@NonNull final int userId) {
                this.userId = userId;
            }
        }
    }
}

File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/repository/retrofit/service/RetrofitGroupService.kt
package app.android.ttp.mikazuki.yoshinani.repository.retrofit.service

import app.android.ttp.mikazuki.yoshinani.model.GroupModel
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.entity.Group
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.entity.GroupUser
import retrofit2.Response
import retrofit2.http.*
import rx.Observable
import java.util.*

/**
 * @author haijimakazuki
 */
interface RetrofitGroupService {

    @GET(PATH_GROUPS)
    fun getAll(@Query("user_id") userId: Int): Observable<Response<List<Group>>>

    @GET(PATH_GROUP_WITH_ID)
    operator fun get(@Path("id") groupId: Int): Observable<Response<Group>>

    @POST(PATH_GROUPS)
    fun create(@Body data: RequestDataOnCreate): Observable<Response<Group>>

    @PATCH(PATH_GROUP_WITH_ID)
    fun update(@Path("id") groupId: Int, @Body data: RequestDataOnCreate): Observable<Response<Group>>

    @POST(PATH_GROUP_USER)
    fun invite(@Path("group_id") groupId: Int,
               @Body data: RequestDataOnInvite): Observable<Response<Group>>

    @PATCH(PATH_GROUP_USER_ACCEPT)
    fun accept(@Path("group_id") groupId: Int): Observable<Response<GroupUser>>

    class RequestDataOnCreate(group: GroupModel) {
        internal var group: PostData

        init {
            this.group = PostData(group.createEntity())
        }

        internal inner class PostData(group: Group) {
            var name: String? = null
            var description: String? = null

            init {
                this.name = group.name
                this.description = group.description
            }
        }
    }

    class RequestDataOnInvite(userId: Int) {
        internal var groupUser: MutableList<PostData>

        init {
            this.groupUser = ArrayList()
            this.groupUser.add(PostData(userId))
        }

        internal inner class PostData(var userId: Int)
    }

    companion object {

        const val PATH_GROUPS = "groups"
        const val PATH_GROUP_WITH_ID = "groups/{id}"
        const val PATH_GROUP_USER = "groups/{group_id}/users"
        const val PATH_GROUP_USER_ACCEPT = "groups/{group_id}/users/accept"
    }
}
